### PR TITLE
Rodeos per block ack

### PR DESCRIPTION
**Pull request by heifner**
## Change Description

- Wait on ack from AMQP before continuing on to processing the next block.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by swatanabe-b1** [comment]
I don't understand the motivation for this change.
- commitTransaction only guarantees that rabbitmq has taken responsibility for the message.  It does not guarantee that the message has been delivered.
- Rodeos can still be ahead of what has been successfully published to rabbit, so it doesn't allow you to get rid of the local storage for messages.

**Comment by heifner** [comment]
> 1. commitTransaction only guarantees that rabbitmq has taken responsibility for the message.  It does not guarantee that the message has been delivered.
> 2. Rodeos can still be ahead of what has been successfully published to rabbit, so it doesn't allow you to get rid of the local storage for messages.

1. The desired responsibility is that rabbitmq has taken responsibility of the message. Once it has ack'ed the message it is rabbitmq responsibility to deliver or recover the message on DR.
2. Rodeos commits the changes to rocksdb after the ack from rabbitmq.

**Comment by swatanabe-b1** [comment]
> 2. Rodeos commits the changes to rocksdb after the ack from rabbitmq.

That's true, but the commit happens regardless of whether the ack was received or not.  The only thing that I see that checks for `app().quit()` is `write_deltas`, which happens before the filter is processed.  `rodeos_db_snapshot::end_block` may commit the changes before we return to appbase and check the quitting flag.

File: programs/rodeos/streams/rabbitmq.hpp [hunk 1]
```diff
@@ -14,7 +14,7 @@
 namespace b1 {
 
 class rabbitmq : public stream_handler {
-   std::shared_ptr<eosio::reliable_amqp_publisher> amqp_publisher_;
+   std::unique_ptr<eosio::reliable_amqp_publisher> amqp_publisher_;
```
Comment (swatanabe-b1) [inline programs/rodeos/streams/rabbitmq.hpp:17]: I don't think it makes sense to use `reliable_amqp_publisher` here anymore.  You're adding workarounds to bypass most of the functionality that it provides.

**Comment by heifner** [comment]
> > 1. Rodeos commits the changes to rocksdb after the ack from rabbitmq.
> 
> That's true, but the commit happens regardless of whether the ack was received or not. The only thing that I see that checks for `app().quit()` is `write_deltas`, which happens before the filter is processed. `rodeos_db_snapshot::end_block` may commit the changes before we return to appbase and check the quitting flag.

Oh, I see what you mean now, thanks.

File: programs/rodeos/streams/rabbitmq.hpp (hunk 1 above)
Comment (heifner) [reply programs/rodeos/streams/rabbitmq.hpp:17]: Yeah. I'll just create a transactional_amqp_publisher roughly based on reliable_amqp_publisher

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include <memory>
+#include <vector>
+#include <functional>
+#include <deque>
+
+namespace eosio {
+
+/*
+ * A transactional_amqp_publisher manages a connection to a single AMQP server that publishes messages to an exchange
+ * in a reliable manner: connections to the AMQP server are retried, publishing occurs in a transaction,
```
Comment (swatanabe-b1) [inline libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:13]: How much sense does it make to retry the connection, when you completely stop rodeos after 15 seconds?  Or asked a different way, why is the timeout so short?  I think it's unlikely that a problem that requires opening a new connection would be resolved that quickly.

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      /// Create a reliable queue to the given server publishing to the given exchange
+      /// \param server_url server url as amqp://...
+      /// \param exchange the exchange to publish to
+      /// \param time_out time to wait on transactional ack from AMQP before calling on_fatal_error
+      /// \param on_fatal_error called from AMQP does not ack transaction in time_out time
+      transactional_amqp_publisher(const std::string& server_url, const std::string& exchange,
+                                   const fc::microseconds& time_out, error_callback_t on_fatal_error);
+
+      /// Publish messages. May be called from any thread except internal thread (do not call from on_fatal_error)
+      /// Blocks until AMQP ack of success transactional commit or on_fatal_error if timeout waiting on ack
```
Comment (swatanabe-b1) [inline libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:38]: This implementation may split the messages into multiple transactions.  The previous implementation using `reliable_amqp_publisher` guaranteed that they were all in the same transaction.  Is that an acceptable change?

Comment (heifner) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:38]: Thanks for pointing that out. I meant to ask @spoonincode about that. Seems like we should just put them all into one transaction unless rabbitmq has something that makes putting into smaller needed.

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp (hunk 2 above)
Comment (heifner) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:13]: During testing I noticed that losing connection to rabbitmq happens fairly often (every couple of hours) and is quickly regained. Something more like 30-60 seconds probably would be better to give it some more time to become available.

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp (hunk 3 above)
Comment (heifner) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:38]: Changed to do all events of a block in one transaction.

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp (hunk 2 above)
Comment (swatanabe-b1) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:13]: Have you not pushed this up yet, or did you change the wrong timeout?

Comment (heifner) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:13]: Looks like I accidentally reverted it when reverting another change to the interface.

File: libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+      /// Create a reliable queue to the given server publishing to the given exchange
+      /// \param server_url server url as amqp://...
+      /// \param exchange the exchange to publish to
+      /// \param time_out time to wait on transactional ack from AMQP before calling on_fatal_error
+      /// \param on_fatal_error called from AMQP does not ack transaction in time_out time
+      transactional_amqp_publisher(const std::string& server_url, const std::string& exchange,
+                                   const fc::microseconds& time_out, error_callback_t on_fatal_error);
+
+      /// Publish messages. May be called from any thread except internal thread (do not call from on_fatal_error)
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/amqp/include/eosio/amqp/transactional_amqp_publisher.hpp:37]: Excellent point. Really you should have one transaction_amqp_publisher per thread.

File: libraries/amqp/transactional_amqp_publisher.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+   ctx.stop();
+   thread.join();
+
+   //if a message is in flight, let it try to run to completion
+   if(in_flight) {
+      auto until = std::chrono::steady_clock::now() + std::chrono::seconds(15);
+      ctx.restart();
+      while(in_flight && ctx.run_one_until(until));
+   }
```
[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply libraries/amqp/transactional_amqp_publisher.cpp:147]: That scenario creates unavoidable races.

[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply libraries/amqp/transactional_amqp_publisher.cpp:147]: There's no way to have such a situation that would not also allow the transaction_amqp_publisher to be called after the destructor finishes, which is definitely UB.

**Review by swatanabe-b1 — APPROVED** [review]

**Pull request merged**